test/uk/gov/hmrc/alcoholdutyreturns/models/returns/AdrReturnSubmissionSpec.scala

```diff
@@ -79,10 +424,840 @@ class AdrReturnSubmissionSpec extends SpecBase {
+        .copy(otherSpiritTypeName = None)
+        .validate()
+        .isValid shouldBe false
+    }
+
+    "fail if grainsQuantities fails to validate" in new SetUp {
+      adrReturnSubmission.spirits.get.spiritsProduced.get
+        .copy(grainsQuantities =
+          adrReturnSubmission.spirits.get.spiritsProduced.get.grainsQuantities.copy(maltedBarley = Some(zero))
+        )
+        .validate()
+        .isValid shouldBe false
+    }
+
+    "fail if hasOtherMaltedGrain, but otherMaltedGrainType is not present" in new SetUp {
+      adrReturnSubmission.spirits.get.spiritsProduced.get
+        .copy(otherMaltedGrainType = None)
+        .validate()
+        .isValid shouldBe false
+    }
+
+    "fail if not hasOtherMaltedGrain, but otherMaltedGrainType is present" in new SetUp {
+      adrReturnSubmission.spirits.get.spiritsProduced.get
+        .copy(hasOtherMaltedGrain = false)
+        .validate()
+        .isValid shouldBe false
+    }
+
+    "fail if ingredientsVolumes fails to validate" in new SetUp {
+      adrReturnSubmission.spirits.get.spiritsProduced.get
+        .copy(ingredientsVolumes =
+          adrReturnSubmission.spirits.get.spiritsProduced.get.ingredientsVolumes.copy(ethylene = Some(zero))
+        )
+        .validate()
+        .isValid shouldBe false
+    }
+
+    "succeed if otherIngredient is not present" in new SetUp {
+      adrReturnSubmission.spirits.get.spiritsProduced.get.copy(otherIngredient = None).validate().isValid shouldBe true
+    }
+
+    "fail if otherIngredient is present but not valid" in new SetUp {
+      adrReturnSubmission.spirits.get.spiritsProduced.get
+        .copy(otherIngredient =
+          Some(adrReturnSubmission.spirits.get.spiritsProduced.get.otherIngredient.get.copy(quantity = zero))
+        )
+        .validate()
+        .isValid shouldBe false
+    }
+  }
+
+  "AdrSpirits" should {
+    "validate if all fields are valid" in new SetUp {
+      adrReturnSubmission.spirits.get.validate(returnPeriod).isValid shouldBe true
+    }
+
+    "fail if declared but not on a quarter" in new SetUp {
+      adrReturnSubmission.spirits.get.validate(notQuarterReturnPeriod).isValid shouldBe false
+    }
+
+    "fail if declared but spiritsProduced is missing" in new SetUp {
+      adrReturnSubmission.spirits.get.copy(spiritsProduced = None).validate(returnPeriod).isValid shouldBe false
+    }
+
+    "fail if declared but spiritsProduced is not valid" in new SetUp {
+      adrReturnSubmission.spirits.get
+        .copy(spiritsProduced =
+          Some(adrReturnSubmission.spirits.get.spiritsProduced.get.copy(otherSpiritTypeName = None))
+        )
+        .validate(returnPeriod)
+        .isValid shouldBe false
+    }
+
+    "fail if not declared but spiritsProduced is present" in new SetUp {
+      adrReturnSubmission.spirits.get.copy(spiritsDeclared = false).validate(returnPeriod).isValid shouldBe false
+    }
+  }
+
+  "AdrTotals" should {
```

Comment:
I am not sure about zero checks for all the fileds in AdrTotals. We made the rates file flexible so theoretically someone can enter a rate band with a rate of 0. But hten the submission will fail here because of a zero check